amazon-kinesis-client/src/test/java/software/amazon/kinesis/leases/LeaseCleanupManagerTest.java

```diff
@@ -0,0 +1,347 @@
+            throws Exception {
+        shardInfo = new ShardInfo("shardId-0", concurrencyToken, Collections.emptySet(),
+                ExtendedSequenceNumber.LATEST);
+
+        verifyExpectedDeletedLeasesCompletedShardCase(shardInfo, childShardsForMerge(), extendedSequenceNumber, 0);
+    }
+
+    /**
+     * Tests that if a lease's parents are still present, we do not delete the lease.
+     */
+    @Test
+    public final void testLeaseNotDeletedWhenParentsStillPresent() throws Exception {
+        shardInfo = new ShardInfo("shardId-0", concurrencyToken, Collections.singleton("parent"),
+                ExtendedSequenceNumber.LATEST);
+
+        verifyExpectedDeletedLeasesCompletedShardCase(shardInfo, childShardsForMerge(), ExtendedSequenceNumber.LATEST, 0);
+    }
+
+    /**
+     * Tests ResourceNotFound case for if a shard expires, that we delete the lease when shardExpired is found.
+     */
+    @Test
+    public final void testLeaseDeletedWhenShardDoesNotExist() throws Exception {
+        shardInfo = new ShardInfo("shardId-0", concurrencyToken, Collections.emptySet(),
+                ExtendedSequenceNumber.LATEST);
+        final Lease heldLease = LeaseHelper.createLease(shardInfo.shardId(), "leaseOwner", Collections.singleton("parentShardId"));
+
+        testLeaseDeletedWhenShardDoesNotExist(heldLease);
+    }
+
+    /**
+     * Tests ResourceNotFound case when completed lease cleanup is disabled.
+     * @throws Exception
+     */
+    @Test
+    public final void testLeaseDeletedWhenShardDoesNotExistAndCleanupCompletedLeaseDisabled() throws Exception {
+        shardInfo = new ShardInfo("shardId-0", concurrencyToken, Collections.emptySet(),
+                ExtendedSequenceNumber.LATEST);
+        final Lease heldLease = LeaseHelper.createLease(shardInfo.shardId(), "leaseOwner", Collections.singleton("parentShardId"));
+
+        cleanupLeasesOfCompletedShards = false;
+
+        leaseCleanupManager = new LeaseCleanupManager(leaseCoordinator, kinesis, NULL_METRICS_FACTORY, maxFutureWait,
+                deletionThreadPool, cleanupLeasesOfCompletedShards, leaseCleanupIntervalMillis, completedLeaseCleanupIntervalMillis,
+                garbageLeaseCleanupIntervalMillis);
+
+        testLeaseDeletedWhenShardDoesNotExist(heldLease);
+    }
+
+    public void testLeaseDeletedWhenShardDoesNotExist(Lease heldLease) throws Exception {
+        when(leaseCoordinator.leaseRefresher()).thenReturn(leaseRefresher);
+        when(leaseCoordinator.getCurrentlyHeldLease(shardInfo.shardId())).thenReturn(heldLease);
+        when(kinesis.getShardIterator(any(GetShardIteratorRequest.class))).thenThrow(ResourceNotFoundException.class);
+        when(leaseRefresher.getLease(heldLease.leaseKey())).thenReturn(heldLease);
+
+        leaseCleanupManager.enqueueForDeletion(new LeasePendingDeletion(streamIdentifier, heldLease, shardInfo));
+        leaseCleanupManager.cleanupLeases();
+
+        verify(leaseRefresher, times(1)).deleteLease(heldLease);
```

Comment:
You can also verify by doing a lookup into the queue

Reply:
Keeping it this way to minimize exposing fields, I don't think other classes should be trying manipulate the queue directly